### PR TITLE
[Inliner] Fix stack nesting when lowering OSSA.

### PR DESCRIPTION
**Pull request by nate-chandler**
Now that in OSSA `partial_apply [on_stack]`s are represented as owned values rather than stack locations, it is possible for their destroys to violate stack discipline.  A direct lowering of the instructions to non-OSSA would violate stack nesting.

Previously, when inlining, it was assumed that non-coroutine callees maintained stack discipline.  And, when inlining an OSSA function into a non-OSSA function, OSSA instructions were lowered directly.  The result was that stack discipline could be violated.

Here, when inlining a function in OSSA form into a function lowered out of OSSA form, stack nesting is fixed up.

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please clean test macOS platform

File: lib/SILOptimizer/Transforms/PerformanceInliner.cpp [hunk 1]
```diff
@@ -1034,10 +1040,30 @@ bool SILPerformanceInliner::inlineCallsIntoFunction(SILFunction *Caller) {
...
     if (SILPrintInliningCallerAfter) {
       printInliningDetailsCallerAfter(PassName, Caller, Callee);
     }
+    if (EnableVerifyAfterEachInlining) {
+      deleter.cleanupDeadInstructions();
+
+      // The inliner splits blocks at call sites. Re-merge trivial branches to
+      // reestablish a canonical CFG.
+      mergeBasicBlocks(Caller);
+
```
Comment (meg-gupta) [inline lib/SILOptimizer/Transforms/PerformanceInliner.cpp:1059]: Is merging basic blocks only required for verification ?

File: lib/SILOptimizer/Mandatory/MandatoryInlining.cpp [hunk 2]
```diff
@@ -972,6 +982,16 @@ runOnFunctionRecursively(SILOptFunctionBuilder &FuncBuilder, SILFunction *F,
       // later.
       changedFunctions.insert(F);
 
+      if (EnableVerifyAfterEachInlining) {
+        if (invalidatedStackNesting) {
+          StackNesting::fixNesting(F);
```
Comment (meg-gupta) [inline lib/SILOptimizer/Mandatory/MandatoryInlining.cpp:987]: This code should be not be covered with `EnableVerifyAfterEachInlining` ?

File: lib/SILOptimizer/Transforms/PerformanceInliner.cpp [hunk 3]
```diff
@@ -1034,10 +1040,30 @@ bool SILPerformanceInliner::inlineCallsIntoFunction(SILFunction *Caller) {
...
+      if (invalidatedStackNesting) {
+        StackNesting::fixNesting(Caller);
```
Comment (meg-gupta) [inline lib/SILOptimizer/Transforms/PerformanceInliner.cpp:1061]: Same as above. `EnableVerifyAfterEachInlining` is covering stack nesting fix.

**Review by meg-gupta — CHANGES_REQUESTED** [review]

**Comment by meg-gupta** [comment]
@nate-chandler CSE calls into inline helper - https://github.com/apple/swift/blob/78749d8d00e5e7fa8c82b6235a10e22c65b01098/lib/SILOptimizer/Transforms/CSE.cpp#L791 You may want to move the stack nest fixup to ` SILInliner::inlineFullApply`

File: lib/SILOptimizer/Transforms/PerformanceInliner.cpp (hunk 1 above)
Comment (nate-chandler) [reply lib/SILOptimizer/Transforms/PerformanceInliner.cpp:1059]: This code runs once after all inlining completes.  Running it after each inlining is only done with `EnableVerifyAfterEachInlining`.

File: lib/SILOptimizer/Transforms/PerformanceInliner.cpp (hunk 3 above)
Comment (nate-chandler) [reply lib/SILOptimizer/Transforms/PerformanceInliner.cpp:1061]: Same as above.

**Comment by nate-chandler** [comment]
@meg-gupta Thanks for pointing that out.  CSE will need to fixup nesting in this case.  We don't want to move it into inlineFullApply because nesting shouldn't be fixed up multiple times when multiple callees are inlined.

File: lib/SILOptimizer/Transforms/PerformanceInliner.cpp (hunk 3 above)
Comment (nate-chandler) [reply lib/SILOptimizer/Transforms/PerformanceInliner.cpp:1061]: And to clarify, the first commit https://github.com/apple/swift/pull/63980/commits/32685ce68e6d38ea43fced760d0fe73806cb73f5 is NFC--it only has any effect when the flag is set.  The actual fixes are the subsequent two commits.

Comment (meg-gupta) [reply lib/SILOptimizer/Transforms/PerformanceInliner.cpp:1061]: Ahh.. makes sense now

**Review by meg-gupta — APPROVED** [review]
@nate-chandler This LGTM.

**Comment by nate-chandler** [comment]
@meg-gupta The CSE fix is here: https://github.com/apple/swift/pull/64005 .

**Comment by nate-chandler** [comment]
@meg-gupta The fix for the last call to inlineFullApply is here: https://github.com/apple/swift/pull/64013 .

**Pull request merged**